plugins/4.applicationInsights.ts

```diff
@@ -0,0 +1,83 @@
+// This plugin was inspired by https://github.com/nuxt-community/applicationinsights-module
+
+/* eslint-disable no-underscore-dangle */
+/* eslint-disable no-param-reassign */
+/* eslint-disable no-nested-ternary */
+import { ApplicationInsights } from "@microsoft/applicationinsights-web";
+
+export default defineNuxtPlugin((nuxtApp) => {
+  const classify = (str: string) =>
+    str.replace(/(?:^|[-_])(\w)/g, (c) => c.toUpperCase()).replace(/[-_]/g, "");
+
+  const formatComponentName = (vm: any, includeFile: boolean) => {
+    if (vm.$root === vm) {
+      return "<Root>";
+    }
+    const options =
+      typeof vm === "function" && vm.cid != null
+        ? vm.options
+        : vm._isVue
+        ? vm.$options || vm.constructor.options
+        : vm || {};
+    let name = options.name || options._componentTag;
+    const file = options.__file;
+    if (!name && file) {
+      const match = file.match(/([^/\\]+)\.vue$/);
+      name = match && match[1];
+    }
+
+    return (
+      (name ? `<${classify(name)}>` : "<Anonymous>") +
+      (file && includeFile !== false ? ` at ${file}` : "")
+    );
+  };
+
+  const setupVueErrorHandling = (vm: any, appInsights: ApplicationInsights) => {
+    const oldErrorHandler = vm.config.errorHandler;
+
+    vm.config.errorHandler = (
+      err: any,
+      context: { $options: { propsData: any } },
+      info: any
+    ) => {
+      appInsights.trackException({
+        exception: err,
+        properties: {
+          errorInfo: info,
+          component: context ? formatComponentName(context, true) : undefined,
+          props: context ? context.$options.propsData : undefined,
+        },
+      });
+      appInsights.flush();
+      if (typeof oldErrorHandler === "function") {
+        oldErrorHandler.call(context, err, context, info);
+      }
+    };
+  };
+
+  const setupPageTracking = (router: any, appInsights: ApplicationInsights) => {
+    router.beforeEach((route: { name: any }, _from: any, next: () => void) => {
+      appInsights.startTrackPage(route.name);
+      next();
+    });
+
+    router.afterEach((route: { name: any }) => {
+      appInsights.stopTrackPage(route.name);
+    });
+  };
+
+  const appInsights = new ApplicationInsights({
+    config: {
+      connectionString:
+        "InstrumentationKey=45d71410-6840-472a-bf95-45f77749c0a2;IngestionEndpoint=https://westeurope-5.in.applicationinsights.azure.com/;LiveEndpoint=https://westeurope.livediagnostics.monitor.azure.com/",
+      /* ...Other Configuration Options... */
+    },
+  });
+  appInsights.loadAppInsights();
+  appInsights.trackPageView();
+
+  inject("appInsights", appInsights);
```

Comment:
What is this `inject('appInsights', appInsights)` ?

Reply:
I assumed that it makes it usable in other components. But I haven't gotten that far yet.
There's a reason it's only a draft PR. I still need to figure out how to provide additional info and how to create custom events.

Reply:
I see! You can return it like this from the plugin:

```ts
return {
	provide: {
		appInsights
	}
}
```

You will then have access to it through the `useNuxtApp()` composable:

```ts
const { $appInsights } = useNuxtApp()
```

Reply:
What's the best way to use appInsights in the 3.mediaPlayer.ts plugin?

Reply:
@kkuepper https://nuxt.com/docs/guide/directory-structure/plugins#typing-plugins - combined with the statement above the anchor, which is what Sigve was pointing out. Auto completion for those helpers is still tricky 🙈

Injecting it the way you did here (https://vuejs.org/guide/components/provide-inject.html) is the Vue-way, but Nuxt has different way, which is the one Sigve mentioned 🤗